rust/mlt-nom/src/v0x01/geometry.rs

```diff
@@ -0,0 +1,198 @@
+        }
+
+        // Use decode_root_length_stream if geometry_offsets is present
+        if let Some(offsets) = geometry_offsets.take() {
+            geometry_offsets = Some(decode_root_length_stream(
+                &vector_types,
+                &offsets,
+                GeometryType::Polygon,
+            ));
+        } else if let Some(offsets) = part_offsets.take() {
+            if let Some(_ring_offsets) = ring_offsets {
+                part_offsets = Some(decode_root_length_stream(
+                    &vector_types,
+                    &offsets,
+                    GeometryType::LineString,
+                ));
+                // decodeLevel1LengthStream(geometryTypes,
+                //                          partOffsets,
+                //                          ringOffsetsCopy,
+                //                          /*isLineStringPresent=*/true,
+                //                          ringOffsets);
+                todo!("part_offsets with ring_offsets case is not implemented");
+            } else {
+                part_offsets = Some(decode_root_length_stream(
+                    &vector_types,
+                    &offsets,
+                    GeometryType::Point,
+                ));
+            }
+        }
+
+        Ok(DecodedGeometry {
+            // vector_type,
+            // vertex_buffer_type: VertexBufferType::Vec2, // Morton not supported yet
+            vector_types,
+            geometry_offsets,
+            part_offsets,
+            ring_offsets,
+            vertex_offsets,
+            index_buffer,
+            triangles,
+            vertices,
+        })
+    }
+}
+
+fn decode_geometry_types(meta: Stream) -> Result<Vec<GeometryType>, MltError> {
+    // TODO: simplify this, e.g. use u8 or even GeometryType directly rather than going via Vec<u32>
+    let vector_types: Vec<u32> = meta.decode_bits_u32()?.decode_u32()?;
+    let vector_types: Vec<GeometryType> = vector_types
+        .into_iter()
+        .map::<Result<GeometryType, MltError>, _>(|v| Ok(u8::try_from(v)?.try_into()?))
+        .collect::<Result<_, _>>()?;
+    Ok(vector_types)
+}
+// pub fn get_vector_type_int_stream(metadata: &Stream) -> VectorType {
+//     match metadata.stream {
+//         StreamType::Rle => {
+//             if metadata.data.len() == 1 {
+//                 VectorType::Const
+//             } else {
+//                 VectorType::Flat
+//             }
+//         }
+//         StreamType::DeltaRle if (1..=2).contains(&metadata.data.len()) => VectorType::Sequence,
+//         _ => {
+//             if metadata.num_values == 1 {
+//                 VectorType::Const
+//             } else {
+//                 VectorType::Flat
+//             }
+//         }
+//     }
+// }
+
+/// Handle the parsing of the different topology length buffers separate not generic to reduce the
+/// branching and improve the performance
```

Comment:
could you reword this?
I don't get what is the generic alternative

Reply:
i think this was copied from the mlt repo - not sure what it means here either - another thing for later rewording